libbpf-rs/src/print.rs

```diff
@@ -0,0 +1,60 @@
+use crate::*;
+use std::os::raw::c_char;
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(u32)]
+pub enum PrintLevel {
+    Warn = libbpf_sys::LIBBPF_WARN,
+    Info = libbpf_sys::LIBBPF_INFO,
+    Debug = libbpf_sys::LIBBPF_DEBUG,
+}
+
+impl From<libbpf_sys::libbpf_print_level> for PrintLevel {
+    fn from(level: libbpf_sys::libbpf_print_level) -> Self {
+        match level {
+            libbpf_sys::LIBBPF_WARN => Self::Warn,
+            libbpf_sys::LIBBPF_INFO => Self::Info,
+            libbpf_sys::LIBBPF_DEBUG => Self::Debug,
+            // shouldn't happen, but anything unknown becomes the highest level
+            _ => Self::Warn,
+        }
+    }
+}
+
+pub type PrintCallback = fn(PrintLevel, &str);
+
+// There is no AtomicFnPtr. This is a workaround.
+static PRINT_CB: AtomicPtr<()> = AtomicPtr::new(std::ptr::null_mut());
+
+/// libbpf's default cb uses vfprintf's return code...which is ignored everywhere. Mimic that for
+/// completeness
+extern "C" fn outer_print_cb(
+    level: libbpf_sys::libbpf_print_level,
+    fmtstr: *const c_char,
+    va_list: *mut libbpf_sys::__va_list_tag,
+) -> i32 {
+    // can't be null: we cant't get here until print_cb has been initialized with a function
+    // pointer.
+    let cb = PRINT_CB.load(Ordering::Relaxed);
+    let cb = unsafe { std::mem::transmute::<*const (), PrintCallback>(cb) };
+    match unsafe { vsprintf::vsprintf(fmtstr, va_list) } {
+        Ok(s) => {
+            cb(level.into(), &s);
+            s.len() as i32
+        }
+        Err(_) => {
+            cb(PrintLevel::Warn, "Could not format libbpf log string");
+            -1
+        }
+    }
+}
+
+/// Set a callback to receive log messages from libbpf, instead of printing them to stderr
+///
+/// This overrides (and is overridden by) [`ObjectBuilder::debug`]
+pub fn set_print(func: PrintCallback) {
+    let cb = unsafe { std::mem::transmute::<PrintCallback, *mut ()>(func) };
+    PRINT_CB.store(cb, Ordering::Relaxed);
+    unsafe { libbpf_sys::libbpf_set_print(Some(outer_print_cb)) };
```

Comment:
libbpf_set_print returns the previous instance of print callback. It's quite important and useful when someone wants to temporarily replace print callback and restore the previous one. Can the same pattern be support in Rust?

Reply:
I can rework this to include that. I was a little unsure about how to handle return values. The return from `libbpf_set_print` would be a nullable, unsafe function pointer, so at best it could be a token to restore things later. Since in my implementation, the real callback is a static function, the `libbpf_set_print`'s return value alone wouldn't be sufficient to restore the print callback in a way the user expects.

Would something like a resource guard be acceptable here, or do you think there is a real need to expose the return value?

```
{
    let _lock = libbpf_rs::set_print_temporary(callback);
    ...
}
// Original callback restored here
```

Reply:
Resource guard seems like an overkill. We can make set_print() not return anything in libbpf-rs. Alternatively you can return PrintCallback and just keep track of current callback on purely Rust side. It might be a bit fragile if someone is calling into libbpf_set_print() through FFI, but that should be extremely rare (an we can say unsupported) case.

Reply:
All right, that should be updated.